File: commons-core/src/test/scala/com/avsystem/commons/serialization/BuildablePojo.java
package com.avsystem.commons.serialization;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public final class BuildablePojo {
    private final String str;
    private final int num;
    private final List<Boolean> flags;
    private final boolean cool;

    private BuildablePojo(String str, int num, List<Boolean> flags, boolean cool) {
        this.str = str;
        this.num = num;
        this.flags = flags;
        this.cool = cool;
    }

    public String getStr() {
        return str;
    }

    public int getNum() {
        return num;
    }

    public List<Boolean> getFlags() {
        return flags;
    }

    public boolean isCool() {
        return cool;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        BuildablePojo that = (BuildablePojo) o;
        return num == that.num && cool == that.cool && Objects.equals(str, that.str) && Objects.equals(flags, that.flags);
    }

    @Override
    public int hashCode() {
        return Objects.hash(str, num, flags, cool);
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private String str = null;
        private int num = 0;
        private List<Boolean> flags = new ArrayList<>();
        private boolean cool = true;

        private Builder() {
        }

        public Builder setStr(String str) {
            this.str = str;
            return this;
        }

        public Builder setNum(int num) {
            this.num = num;
            return this;
        }

        public Builder setFlags(List<Boolean> flags) {
            this.flags = flags;
            return this;
        }

        public Builder setCool(boolean cool) {
            this.cool = cool;
            return this;
        }

        public BuildablePojo build() {
            return new BuildablePojo(str, num, flags, cool);
        }
    }
}

File: core/src/test/scala/com/avsystem/commons/serialization/BuildablePojo.scala
package com.avsystem.commons
package serialization

// this should be written in Java but it was rewritten to Scala in order to make Scala.js linker happy
object BuildablePojo {
  def builder(): Builder = new Builder

  final class Builder private[BuildablePojo] {
    private var str: String = scala.compiletime.uninitialized
    private var num: Int = 0
    private var flags: JList[Boolean] = new JArrayList
    private var cool: Boolean = true

    def setStr(str: String): Builder = {
      this.str = str
      this
    }

    def setNum(num: Int): Builder = {
      this.num = num
      this
    }

    def setFlags(flags: JList[Boolean]): Builder = {
      this.flags = flags
      this
    }

    def setCool(cool: Boolean): Builder = {
      this.cool = cool
      this
    }

    def build(): BuildablePojo =
      new BuildablePojo(str, num, flags, cool)
  }
}
final class BuildablePojo private (
  private val str: String,
  private val num: Int,
  private val flags: JList[Boolean],
  private val cool: Boolean,
) {
  def getStr(): String = str
  def getNum(): Int = num
  def getFlags(): JList[Boolean] = flags
  def isCool(): Boolean = cool

  override def equals(o: Any): Boolean = o match {
    case bp: BuildablePojo =>
      bp.str == str && bp.num == num && bp.flags == flags && bp.cool == cool
    case _ => false
  }

  override def hashCode: Int = (str, num, flags, cool).hashCode

  override def toString: String =
    s"BuildablePojo($str, $num, $flags, $cool)"
}
